src/conn/conn_chunkcache.c

```diff
@@ -0,0 +1,346 @@
+            WT_ERR_NOTFOUND_OK(__chunkcache_metadata_delete(cursor, entry), false);
+        else {
+            __wt_verbose_error(
+              session, WT_VERB_CHUNKCACHE, "got invalid event type %d\n", entry->type);
+            ret = -1;
+            goto err;
+        }
+
+        __wt_free(session, entry);
+        entry = NULL;
+    }
+
+err:
+    if (cursor != NULL)
+        WT_TRET(cursor->close(cursor));
+    if (entry != NULL)
+        __wt_free(session, entry);
+    return (ret);
+}
+
+/*
+ * __chunkcache_metadata_server --
+ *     Dispatch chunks of work (or stop the server) whenever we're signalled to do so.
+ */
+static WT_THREAD_RET
+__chunkcache_metadata_server(void *arg)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    WT_SESSION_IMPL *session;
+    uint64_t cond_time_us;
+    bool signalled;
+
+    session = arg;
+    conn = S2C(session);
+    cond_time_us = 1 * WT_MILLION;
+
+    for (;;) {
+        /* Wait until the next event. */
+        __wt_cond_wait_signal(session, conn->chunkcache_metadata_cond, cond_time_us,
+          __chunkcache_metadata_run_chk, &signalled);
+
+        if (!__chunkcache_metadata_run_chk(session))
+            break;
+
+        if (signalled)
+            WT_ERR(__chunkcache_metadata_work(session));
+    }
+
+    if (0) {
+err:
+        WT_IGNORE_RET(__wt_panic(session, ret, "%s", "chunk cache metadata server error"));
+    }
+
+    return (WT_THREAD_RET_VALUE);
+}
+
+/*
+ * __wt_chunkcache_metadata_create --
+ *     Start the server component of the chunk cache metadata subsystem.
+ */
+int
+__wt_chunkcache_metadata_create(WT_SESSION_IMPL *session)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_RET;
+    char *metadata_config;
+
+    conn = S2C(session);
+    chunkcache = &conn->chunkcache;
+
+    if (!F_ISSET(chunkcache, WT_CHUNKCACHE_CONFIGURED) || chunkcache->type != WT_CHUNKCACHE_FILE)
+        return (0);
+
+    /* Retrieve the chunk cache metadata config, and ensure it matches our startup config. */
+    ret = __chunkcache_get_metadata_config(session, &metadata_config);
+    if (ret == WT_NOTFOUND) {
+        WT_RET(__chunkcache_create_metadata_file(
+          session, chunkcache->capacity, chunkcache->hashtable_size, chunkcache->chunk_size));
+        __wt_verbose(session, WT_VERB_CHUNKCACHE, "%s", "created chunkcache metadata file");
+    } else if (ret == 0) {
+        WT_RET(__chunkcache_verify_metadata_config(session, metadata_config, chunkcache->capacity,
+          chunkcache->hashtable_size, chunkcache->chunk_size));
+        __wt_verbose(session, WT_VERB_CHUNKCACHE, "%s", "reused chunkcache metadata file");
+    } else {
+        WT_RET(ret);
+    }
+    __wt_free(session, metadata_config);
+
+    /* Start the internal thread. */
+    WT_ERR(__wt_cond_alloc(session, "chunkcache metadata", &conn->chunkcache_metadata_cond));
+    FLD_SET(conn->server_flags, WT_CONN_SERVER_CHUNKCACHE_METADATA);
+
+    /* TODO set isolation */
```

Comment:
This bit could use some advice. I don't hate that it's defaulting to snapshot isolation, but it's much stronger than we need. Unfortunately, setting it to read uncommitted (or even read committed) leads to the assertion associated with this failing:

```
    /*
     * Return error when the transactions with read committed or uncommitted isolation tries to
     * perform any write operation. Don't return an error for any update on metadata because it uses
     * special transaction visibility rules, search and updates on metadata happens in
     * read-uncommitted and read-committed isolation.
     */
```

It sounds deliberate that we can't use a lower isolation level for an internal session - can anyone confirm/deny?

Reply:
Closing this out - lower isolation levels are deliberately read-only, and so are incompatible with this use case.